activerecord/lib/active_record/relation/predicate_builder/association_query_value.rb

```diff
@@ -9,7 +9,7 @@ def initialize(associated_table, value)
       end
 
       def queries
-        [ associated_table.join_foreign_key => ids ]
+        ids.map { |ids_set| Array(associated_table.join_foreign_key).zip(Array(ids_set)).to_h }
```

Comment:
Let me know if you would prefer branching to avoid unnecessary array allocations from `Array()` for regular models. 

Under the hood for the models with composite `query_constraints` the resulting value looks like
```ruby
[{blog_id: 1, id: 1}, {blog_id: 2, id: 2}]
```
Which later gets translated into a series of `OR` statements which is what we expect.
https://github.com/rails/rails/blob/0020330c77a46be245369dc65d93196ad54d28ac/activerecord/lib/active_record/relation/predicate_builder.rb#L140

Reply:
Yes let's branch for a couple reasons:

1) Doesn't slow down the default path
2) Makes it easier for us to find the patterns we want to replace after the features are set.

Reply:
Done. I end up branching on `if associated_table.join_foreign_key.is_a?(Array)` even though our preference is to branch on `klass.query_constraints_list` presence however in this context we don't have access to a `klass` (`Sharded::BlogPost`) in our case. 
It could have been possible to access them through `associated_table.klass` or through looking it up in `associated_table.reflection...` but both of those values are private to `associated_table`
https://github.com/rails/rails/blob/6854a315aa87953e9d15adbe6d46f1953ac93238/activerecord/lib/active_record/table_metadata.rb#L87-L88

I don't feel confident making either of these public as our ultimate goal is to be able to get rid of branching which means changing api of `TableMetadata` won't be aligned with the long-term vision.

I should additionally mention that even with a `klass` being available, we shouldn't be using `klass.composite_primary_key?` as `composite_primary_key?` only relevant on the lower db-level context while we want to change behavior for the virtual `query_constraints` use-case